### PR TITLE
feat(mito): limit compaction output file size

### PR DESCRIPTION
**Pull request by v0y4g3r**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

This PR adds a limit to the compaction output file size so that compaction will not generate a huge file which may slow down queries.

Now we can create tables with explicit output file limit:
```sql
CREATE TABLE monitor (
    host STRING,
    ts TimestampSecond DEFAULT CURRENT_TIMESTAMP() TIME INDEX,
    cpu FLOAT64 DEFAULT 0,
    memory FLOAT64,
    PRIMARY KEY(host)
) WITH (
    'append_mode'='true', 
    'compaction.type'='twcs', 
    'compaction.twcs.max_output_file_size'='1GB'
);
```

SST files distribution under TSBS workload (with max_output_file_size=128MB):
```
 # Before
-rw-rw-r-- 1 lei lei 135M Sep 20 16:38 32f86803-e73e-4ca2-a61e-68dc45b653e0.parquet
-rw-rw-r-- 1 lei lei  46M Sep 20 16:36 69539117-4acc-4f8f-9af2-e685c505c5b2.parquet
-rw-rw-r-- 1 lei lei  91M Sep 20 16:37 838dfa1c-1330-499d-8a15-5a0179ce801e.parquet
-rw-rw-r-- 1 lei lei  46M Sep 20 16:38 9cd89ebb-044a-450b-98e2-ab3ffa40e806.parquet
-rw-rw-r-- 1 lei lei 279M Sep 20 16:37 9d8e7106-62b5-4a72-ab4e-684a246fd326.parquet
-rw-rw-r-- 1 lei lei 273M Sep 20 16:35 af108633-fbc1-44f7-a352-7be08431cd97.parquet

# After
-rw-rw-r-- 1 lei lei  90M Sep 20 22:19 0e172b52-ebf8-4ac0-b1f3-34f17154295e.parquet
-rw-rw-r-- 1 lei lei  91M Sep 20 22:18 1f2ac123-f381-4cf6-b0f7-760f6ceda18a.parquet
-rw-rw-r-- 1 lei lei  90M Sep 20 22:19 3c47a5dd-c648-4b61-aa48-ae0f55bb8686.parquet
-rw-rw-r-- 1 lei lei  90M Sep 20 22:16 3e7b24dd-cfe3-4155-a268-5f2dba6d2bd3.parquet
-rw-rw-r-- 1 lei lei  90M Sep 20 22:19 4bbbcd26-fd9a-444b-b14d-db11e20b1350.parquet
-rw-rw-r-- 1 lei lei  90M Sep 20 22:17 8432536a-93e3-45fa-ab8f-f63ad1d90fa1.parquet
-rw-rw-r-- 1 lei lei  46M Sep 20 22:19 c66a2604-b076-4eec-b085-2a013f5ff6c0.parquet
-rw-rw-r-- 1 lei lei  91M Sep 20 22:16 cb2dbe3f-691b-4dc9-a877-7213282905b4.parquet
-rw-rw-r-- 1 lei lei  91M Sep 20 22:18 d5b72303-7fbe-4aef-904e-6e681b68f0bb.parquet
-rw-rw-r-- 1 lei lei  91M Sep 20 22:16 fa146609-2875-44f9-9a7c-094e968d5aa4.parquet
```
## Caveats

Currently this policy only applies when either append mode is enabled or compacted inputs does not overlap with other inputs (so that we can reasobably estimate the ouput file size). For those overlapping files, we may need to examine row groups' statistic data to further calculate the output file size. But this is not oncluded in this patch.

## Checklist

- [x] I have written the necessary rustdoc comments.
- [x] I have added the necessary unit tests and integration tests.
- [x] This PR requires documentation updates.

### CODE REVIEW

File: src/mito2/src/compaction/twcs.rs [hunk 1]
```diff
@@ -35,30 +35,23 @@ const LEVEL_COMPACTED: Level = 1;
...
+    pub max_active_window_files: usize,
+    /// Max allowed sorted runs in inactive windows.
+    pub max_inactive_window_runs: usize,
+    /// Max allowed files in inactive windows.
+    pub max_inactive_window_files: usize,
+    /// Compaction time window in seconds.
+    pub time_window_seconds: Option<i64>,
+    /// Max allowed compaction output file size.
+    pub max_output_file_size: Option<u64>,
+    // Whether the target region is in append mode.
```
Comment (waynexia) [inline src/mito2/src/compaction/twcs.rs:50]:
```suggestion
    /// Whether the target region is in append mode.
```

**Review by waynexia — COMMENTED** [review]
what happens if those params are in conflict and cannot all be met?

File: src/mito2/src/compaction/twcs.rs [hunk 2]
```diff
@@ -84,37 +77,90 @@ impl TwcsPicker {
 
             // we only remove deletion markers once no file in current window overlaps with any other window.
             let found_runs = sorted_runs.len();
-            let filter_deleted = !files.overlapping && (found_runs == 1 || max_runs == 1);
+            let mut filter_deleted = !files.overlapping && (found_runs == 1 || max_runs == 1);
```
Comment (evenyag) [inline src/mito2/src/compaction/twcs.rs:80]: Consider assigning this flag in one line.
```suggestion
            let filter_deleted = !files.overlapping && (found_runs == 1 || max_runs == 1) && !self.append_mode;
```

File: src/mito2/src/compaction/twcs.rs [hunk 3]
```diff
@@ -84,37 +77,90 @@ impl TwcsPicker {
...
+                for inputs in maybe_enforce_max_output_size {
+                    if inputs.len() > 1 {
+                        output.push(CompactionOutput {
+                            output_file_id: FileId::random(),
+                            output_level: LEVEL_COMPACTED, // always compact to l1
+                            inputs,
+                            filter_deleted,
+                            output_time_range: None, // we do not enforce output time range in twcs compactions.
+                        });
+                    }
```
Comment (evenyag) [inline src/mito2/src/compaction/twcs.rs:128]: The logic is similar to the logic in the `else if files.files.len() > max_files`. Can we reuse the code? We can invoke `reduce_runs()` or `enforce_file_num` in different branches to get `files_to_compact`.
```rust
let files_to_compact = if found_runs > max_runs {
    let files_to_compact = reduce_runs(sorted_runs, max_runs);
    // ...
    files_to_compact
} else if files.files.len() > max_files {
    let to_merge = enforce_file_num(&files.files, max_files);
    // ...
    vec![to_merge]
} else {
    return;
}

let inputs = if !filter_deleted && let Some(max_output_file_size) = self.max_output_file_size {
    enforce_max_output_size()
} else {
    inputs
};
for input in inputs {
    // ...
}
```

**Comment by v0y4g3r** [comment]
> what happens if those params are in conflict and cannot all be met?

`max_output_file_size` has higher priority than `max_[in]active_window_files`. For example, if `max_[in]active_window_files` allows at most 1 file per window, while the output file size exceeds`max_output_file_size`, the output will be split.

**Comment by v0y4g3r** [comment]
> We should allow a compaction task to generate multiple files to avoid outputting a huge file. So the task can finish writing the current output file and start a new output file. But this requires the compaction task itself to generate file ids.

This will not be included in this PR since it involves refactoring both `WriteCache` and `ParquetWriter`.

File: src/mito2/src/compaction/twcs.rs (hunk 3 above)
Comment (v0y4g3r) [reply src/mito2/src/compaction/twcs.rs]: > The logic is similar to the logic in the `else if files.files.len() > max_files`. Can we reuse the code? We can invoke `reduce_runs()` or `enforce_file_num` in different branches to get `files_to_compact`.
> 
> ```rust
> let files_to_compact = if found_runs > max_runs {
>     let files_to_compact = reduce_runs(sorted_runs, max_runs);
>     // ...
>     files_to_compact
> } else if files.files.len() > max_files {
>     let to_merge = enforce_file_num(&files.files, max_files);
>     // ...
>     vec![to_merge]
> } else {
>     return;
> }
> 
> let inputs = if !filter_deleted && let Some(max_output_file_size) = self.max_output_file_size {
>     enforce_max_output_size()
> } else {
>     inputs
> };
> for input in inputs {
>     // ...
> }
> ```

fixed in 2aa6d4263a

File: src/mito2/src/compaction/twcs.rs (hunk 2 above)
Comment (v0y4g3r) [reply src/mito2/src/compaction/twcs.rs]: > Consider assigning this flag in one line.

fixed in 2aa6d4263a

File: src/mito2/src/compaction/twcs.rs (hunk 1 above)
Comment (v0y4g3r) [reply src/mito2/src/compaction/twcs.rs]: > 

fixed in 2aa6d4263a

**Review by waynexia — APPROVED** [review]
It might also be beneficial to group small size inputs into a larger one while still keeping the total size under the threshold. But this PR looks fine to merge and we can do this in follow-up changes.

**Review by evenyag — APPROVED** [review]

**Pull request merged**